src/Workspaces/Core/Portable/Remote/RemoteHostClientExtensions.cs

```diff
@@ -1,36 +1,100 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Internal.Log;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
     internal static class RemoteHostClientExtensions
     {
-        [Obsolete]
-        public static Task<RemoteHostClient.Session> CreateCodeAnalysisServiceSessionAsync(
-            this RemoteHostClient client, Solution solution, CancellationToken cancellationToken)
+        /// <summary>
+        /// Create <see cref="RemoteHostClient.Session"/> for the <paramref name="serviceName"/> if possible.
+        /// otherwise, return null.
+        /// 
+        /// Creating session could fail if remote host is not available. one of example will be user killing
+        /// remote host.
+        /// </summary>
+        public static Task<RemoteHostClient.Session> TryCreateServiceSessionAsync(this RemoteHostClient client, string serviceName, CancellationToken cancellationToken)
         {
-            return CreateCodeAnalysisServiceSessionAsync(
-                client, solution, callbackTarget: null, cancellationToken: cancellationToken);
+            return client.TryCreateSessionAsync(serviceName, callbackTarget: null, cancellationToken: cancellationToken);
         }
 
-        [Obsolete]
-        public static Task<RemoteHostClient.Session> CreateCodeAnalysisServiceSessionAsync(
-            this RemoteHostClient client, Solution solution, object callbackTarget, CancellationToken cancellationToken)
+        /// <summary>
+        /// Create <see cref="SolutionAndSessionHolder"/> for the <paramref name="serviceName"/> if possible.
+        /// otherwise, return null.
+        /// 
+        /// Creating session could fail if remote host is not available. one of example will be user killing
+        /// remote host.
+        /// </summary>
+        public static Task<SolutionAndSessionHolder> TryCreateServiceSessionAsync(this RemoteHostClient client, string serviceName, Solution solution, CancellationToken cancellationToken)
```

Comment:
It feels weird that this type would leak out.  Why does the caller (i.e. a feature) need to get this type, instead of just getting a regular session that tehy can use?

--

Overall, it feels weird to me that all these new types have been introduced *and* made part of the public API for OOP that features will use.

From my own experience, pretty much all our features want to do is get a session and invoke something on that session.  Everything else is an internal impl detail.